### PR TITLE
Moving updated codebase into main branch ready for environment testing

### PR DESCRIPTION
**Pull request by Ngriffiths1**
Moved updated codebase from POC branch with all changes that were suggested. 
This is part of the API filing project from ticket FA-861 
subtask FA-1005

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,45 +1,107 @@
...
+    </dependency>
+
+    <dependency>
+      <artifactId>validation-api</artifactId>
+      <groupId>javax.validation</groupId>
+      <version>2.0.0.Final</version>
+    </dependency>
+
+  </dependencies>
+  <description>Third party web application</description>
```
Comment (divyaparameswaran) [inline pom.xml]: Can you please move the description , groupId model and name on top ?

Comment (divyaparameswaran) [inline pom.xml]: Can you please move the description , groupId model and name on top ?

File: pom.xml [hunk 2]
```diff
@@ -1,45 +1,107 @@
...
+  <groupId>uk.gov.companieshouse</groupId>
+  <modelVersion>4.0.0</modelVersion>
+
+  <name>third-party-test-harness</name>
+
+  <parent>
```
Comment (divyaparameswaran) [inline pom.xml]: Can you move the parent on top as well please ?

File: pom.xml [hunk 3]
```diff
@@ -1,45 +1,107 @@
...
+    <artifactId>spring-boot-starter-parent</artifactId>
+    <groupId>org.springframework.boot</groupId>
+    <relativePath/>
+    <version>2.3.5.RELEASE</version> <!-- lookup parent from repository -->
+  </parent>
+
+  <properties>
```
Comment (divyaparameswaran) [inline pom.xml]: Properties and version should go on top as well

File: pom.xml [hunk 4]
```diff
@@ -1,45 +1,106 @@
...
+      <plugin>
+        <artifactId>spring-boot-maven-plugin</artifactId>
         <groupId>org.springframework.boot</groupId>
-        <artifactId>spring-boot-starter-parent</artifactId>
-        <version>2.3.5.RELEASE</version>
-        <relativePath/> <!-- lookup parent from repository -->
-    </parent>
+      </plugin>
+      <plugin>
+        <groupId>com.google.cloud.tools</groupId>
```
Comment (divyaparameswaran) [inline pom.xml:33]: Please update the Readme with a section on how to run in docker.

Comment (Ngriffiths1) [reply pom.xml:33]: @divya051988 We will have to wait for @gevans-ch to be back in Monday to write this up in the ReadMe as he completed the work for it

File: src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.time.Instant;
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java]: Please remove unused import

File: src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java]: Same as above

File: src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        when(responseEntity.getBody()).thenReturn(returnTokenString);
+
+        String accessToken = service.getAccessToken(authCode);
+        assertEquals(expectedAccessToken, accessToken);
+
+        ArgumentCaptor<HttpEntity<String>> entityCaptor = ArgumentCaptor.forClass(HttpEntity.class);
+        verify(restTemplate).exchange(eq(accessTokenUrl), eq(HttpMethod.POST), entityCaptor.capture(), eq(String.class));
+
+        HttpEntity<String> entity = entityCaptor.getValue();
+        assertTrue(entity.getHeaders().containsKey("Authorization"));
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java]: The string `Authorization` can be added to a constant

File: src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        assertEquals(1, entity.getHeaders().get("Authorization").size());
+        assertEquals("Basic " + expectedEncodedCredentials, entity.getHeaders().get("Authorization").get(0));
+
+        assertTrue(entity.getHeaders().containsKey("Accept"));
+        assertEquals(1, entity.getHeaders().get("Accept").size());
+        assertEquals("application/json", entity.getHeaders().get("Accept").get(0));
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java]: `Accept` can also go in a constant

File: src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        String accessToken = "PPAK";
+        String email = "email@domain.com";
+        service.storeUserDetails(email, accessToken);
+    
+        assertTrue(service.userTokenStoreMap.containsKey(email));
+        UserTokenStore userToken = service.userTokenStoreMap.get(email);
+        assertEquals(email, userToken.getEmail());
+        assertEquals(accessToken, userToken.getAccessToken());
+    }
+}
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/service/impl/UserAuthServiceImplTest.java:124]: Could you please add unit tests for Controllers as well .

File: src/test/java/uk/gov/companieshouse/controllers/ThirdPartyControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    }
+
+    @Test
+    @DisplayName("Test that the redirect page is reached after interaction with account")
+    void testGetSuccessfulRedirectPage() {
+
+    }
+
+}
```
Comment (divyaparameswaran) [inline src/test/java/uk/gov/companieshouse/controllers/ThirdPartyControllerTest.java:85]: Not sure if this is a typo or the test are supposed to be empty ?

File: pom.xml (hunk 4 above)
Comment (gevans-ch) [reply pom.xml:33]: Updated readme with Docker notes

**Comment by divyaparameswaran** [comment]
@rgoswami-ch Could you please have a look at the comments

**Review by divyaparameswaran — APPROVED** [review]

**Pull request merged**